### PR TITLE
[WIP] rebase: Call ProcessNewBlock() asynchronously

### PR DESCRIPTION
**Pull request by dongcarl**
This is a (currently naive) rebase of #16323, which is a rework of #16175, which is a rework of #12934.
Built on top of #17479.

Currently `validationinterface_tests/unregister_all_during_call` fails.

-----

Goals:
- [ ] Add as much documentation as possible to aid with review
- [ ] Split up commits as much as possible to aid with review

### CODE REVIEW

**Comment by dongcarl** [comment]
Since this PR touches critical parts of the codebase, I think it deserves a thorough review. I'm going to go through the original PR (#16323) commit by commit and write down my notes here: https://docs.google.com/document/d/1tduRmqcvhdl3FRkmdnj0f3_fknXuIZ57R8zdCGHNt6k/edit?usp=sharing

**Comment by dongcarl** [comment]
Rebased on newer #17479. `git-bisect`ing the `validationinterface_tests/unregister_all_during_call` says:

```
194935b1a2968b594a42cb880e30701dd2e2bc7c is the first bad commit
```

**Comment by ryanofsky** [comment]
> Rebased on newer #17479. `git-bisect`ing the `validationinterface_tests/unregister_all_during_call` says:
> 
> ```
> 194935b1a2968b594a42cb880e30701dd2e2bc7c is the first bad commit
> ```

Test was relying on the fact that BlockChecked as synchronous and that commit made it asynchronous. You can switch it to a different synchronous method:

```diff
--- a/src/test/validationinterface_tests.cpp
+++ b/src/test/validationinterface_tests.cpp
@@ -57,15 +57,13 @@ public:
     {
         if (m_on_destroy) m_on_destroy();
     }
-    void BlockChecked(const CBlock& block, const BlockValidationState& state) override
+    void NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock> &block) override
     {
         if (m_on_call) m_on_call();
     }
     static void Call()
     {
-        std::shared_ptr<const CBlock> block = std::make_shared<CBlock>();
-        BlockValidationState state;
-        GetMainSignals().BlockChecked(block, state);
+        GetMainSignals().NewPoWValidBlock(nullptr, std::make_shared<CBlock>());
     }
     std::function<void()> m_on_call;
     std::function<void()> m_on_destroy;
```

**Comment by dongcarl** [comment]
Now `p2p_unrequested_blocks.py` is failing. `git-bisect` says: 
```
1d9f66ea37ac6f22f26013cd57ed75fc04a9481f is the first bad commit
```
-----

As noted in the Google Doc:

> Up to this point, it seems that splitting this commit into:
> 
> 1. Changing the call semantics of `ProcessNewBlock`
> 2. Changing the return type to a `std::future`
> 
> Might make it easier to review.

So I plan to split it up and track it down.

**Comment by dongcarl** [comment]
Split up the first commit into:

1. cb178903b0752de3edfd87f6013c2b4de41b6e13 which changes the call semantics of `ProcessNewBlock` and deals with the fallout of that
2. 7532cf6c56d015e8bae646138d71df825d2684ab which changes the return from a bool to a future of a bool and deals with the fallout of that
3. 2dcdb537b5c1dce60db217238c5e75dae602e755 which contains changes that are non-obvious to me to be correct

-----

I also botched the `git-bisect` before (didn't call `make` in a `bash -c`), new output:
```
$ git bisect run bash -c "make -j50 && python3 test/functional/p2p_unrequested_blocks.py"

...

7c77827558715180594f5881b6d02982504c4fad is the first bad commit
commit 7c77827558715180594f5881b6d02982504c4fad
Author: Matt Corallo <git@bluematt.me>
Date:   Mon Jun 17 13:13:36 2019 -0400

    Move net_processing's ProcessNewBlock calls to resolve async.

    Essentially, our goal is to not process anything for the given peer
    until the block finishes processing (emulating the previous behavior)
    without actually blocking the ProcessMessages loops. Obviously, in
    most cases, we'll just go on to the next peer and immediately hit a
    cs_main lock, blocking us anyway, but this we can slowly improve
    that state over time by moving things from CNodeState to CPeerState.

 src/net_processing.cpp | 79 ++++++++++++++++++++++++++++++++++++++++++--------
 1 file changed, 67 insertions(+), 12 deletions(-)
bisect run success
```

**Comment by dongcarl** [comment]
Got the unit tests and functional tests to pass!

File: src/net_processing.cpp [hunk 1]
```diff
@@ -237,6 +248,30 @@ namespace {
 } // namespace
 
 namespace {
+/**
+ * Maintain state about nodes, protected by our own lock. Historically we put all
+ * peer tracking state in CNodeState, however this results in significant cs_main
+ * contention. Thus, new state tracking should go here, and we should eventually
+ * move most (non-validation-specific) state here.
```
Comment (ariard) [inline src/net_processing.cpp:255]: I think you should define more difference between validation-state and non-validation-state. Right now peer tracking state is spread among multiple class like `CNode` with `TxRelay` or `CNodeState` and now `CPeerState`. We should have a clear idea of _what_ should go _where_, according to which thread uses it. You should also explain how `CPeerState` aims to reduce `cs_main` contention with regards to the new threading model.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -237,6 +248,30 @@ namespace {
...
+ */
+struct CPeerState {
+    //! If this peer generated some headers for us to add, we store the resulting
+    //! future here and wait for it to complete before we process more data from this
+    //! peer.
+    std::future<bool> pending_block_processing;
+    //! The hash of the block which is pending download.
+    uint256 pending_block_hash;
+    //! Once we've finished processing a block from this peer, we must still wait for
+    //! any related callbacks to fire (to ensure, specifically, that rejects go out
```
Comment (ariard) [inline src/net_processing.cpp:265]: Is reject here making reference to reject messages ? I think it doesn't make sense anymore post-#15437 and post-#17004. Also you may lay out the expected callback sequence (as we do for `TransactionRemovedFromMempool` in `src/validationinterface.h`).

File: src/net_processing.cpp [hunk 3]
```diff
@@ -1973,6 +2047,23 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
     }
 }
 
+/**
+ *  A block has been processed. Handle potential peer punishment and housekeeping.
```
Comment (ariard) [inline src/net_processing.cpp:2051]: Please define more "housekeeping".

File: src/net_processing.cpp [hunk 4]
```diff
@@ -3082,30 +3167,22 @@ bool ProcessMessage(CNode* pfrom, const std::string& msg_type, CDataStream& vRec
...
-                // between here and cs_main in ProcessNewBlock is fine.
-                // BIP 152 permits peers to relay compact blocks after validating
-                // the header only; we should not punish peers if the block turns
-                // out to be invalid.
-                mapBlockSource.emplace(resp.blockhash, std::make_pair(pfrom->GetId(), false));
             }
         } // Don't hold cs_main when we call into ProcessNewBlock
         if (fBlockRead) {
-            bool fNewBlock = false;
+            // BIP 152 permits peers to relay compact blocks after validating
```
Comment (ariard) [inline src/net_processing.cpp:3173]: I'm not sure that BIP152 low-bandwidth authorizes invalid header propagation, maybe precise.

File: src/validation.cpp [hunk 5]
```diff
@@ -3729,23 +3729,10 @@ static FlatFilePos SaveBlockToDisk(const CBlock& block, int nHeight, const CChai
     return blockPos;
 }
 
-/** Store block on disk. If dbp is non-nullptr, the file is known to already reside on disk */
-bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, BlockValidationState& state, const CChainParams& chainparams, CBlockIndex** ppindex, bool fRequested, const FlatFilePos* dbp, bool* fNewBlock)
+bool CChainState::ShouldMaybeWrite(CBlockIndex* pindex, bool fRequested)
```
Comment (ariard) [inline src/validation.cpp:3732]: Please can you comment on what conditional write is laying on.

File: src/validation.cpp [hunk 6]
```diff
@@ -3812,39 +3844,135 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, Block
...
+
+            if (state.IsError()) {
+                result_promise.set_value(false);
+                continue;
+            }
+        }
+
+        NotifyHeaderTip();
+
+        BlockValidationState state; // Only used to report errors, not invalidity - ignore it
```
Comment (ariard) [inline src/validation.cpp:3926]: You should point where errors are distinguished from invalidity.

**Review by ariard — COMMENTED** [review]
Great to see progress on this, for now just high-level comments calling for better comments . I will review John PR again and then your linked notes to help to make them easier to read. I think that's the right way to come with a clear display of what is the current validation model, what this changes propose to do and why. There is so much context, that we can't expect all reviewers to do the PR context history digging by themselves.

**Comment by dongcarl** [comment]
Not planning to work on this anytime soon unfortunately.

**Comment by jnewbery** [comment]
image

🙁 

Let me know if you pick this up again!

**Comment by hebasto** [comment]
From my research it follows that the "[Move BlockChecked to a background thread](https://github.com/bitcoin/bitcoin/pull/18963/commits/2ed0b647435b80946f7f000d573ff4b4e274ca70)" commit fixes a lock-order-inversion in the MessageHandler thread.

See https://github.com/bitcoin/bitcoin/issues/19303#issuecomment-1514928912.

**Pull request closed without merging**